runtime/lua/vim/_editor.lua

```diff
@@ -1036,56 +1036,97 @@ end
-    major, minor = tonumber(major), tonumber(minor)
-
-    local hard_deprecated_since = string.format('nvim-%d.%d', major, minor - 1)
-    -- Assume there will be no next minor version before bumping up the major version
-    local is_hard_deprecated = minor == 0 or vim.fn.has(hard_deprecated_since) == 1
-    if not is_hard_deprecated then
-      return
-    end
-
-    local removal_version = string.format('nvim-%d.%d', major, minor)
-    will_be_removed = vim.fn.has(removal_version) == 1 and 'was removed' or will_be_removed
-  else
-    vim.validate {
-      name = { name, 'string' },
-      alternative = { alternative, 'string', true },
-      version = { version, 'string', true },
-      plugin = { plugin, 'string', true },
-    }
-  end
+  vim.validate {
+    name = { name, 'string' },
+    alternative = { alternative, 'string', true },
+    version = { version, 'string', true },
+    plugin = { plugin, 'string', true },
+  }
 
   local msg = ('%s is deprecated'):format(name)
   msg = alternative and ('%s, use %s instead.'):format(msg, alternative) or (msg .. '.')
-  msg = ('%s%s\nFeature %s in %s %s'):format(
-    msg,
-    (plugin == 'Nvim' and ' :help deprecated' or ''),
-    will_be_removed,
-    plugin,
-    version
-  )
+  msg = ('%s\nFeature will be removed in %s %s'):format(msg, plugin, version)
   local displayed = vim.notify_once(msg, vim.log.levels.WARN)
   if displayed and backtrace ~= false then
     vim.notify(debug.traceback('', 2):sub(2), vim.log.levels.WARN)
   end
   return displayed and msg or nil
 end
 
+do
+  local function truncated_echo(msg)
+    -- Truncate message to avoid hit-enter-prompt
+    local max_width = vim.o.columns * math.max(vim.o.cmdheight - 1, 0) + vim.v.echospace
+    local msg_truncated = string.sub(msg, 1, max_width)
+    vim.api.nvim_echo({ { msg_truncated, 'WarningMsg' } }, true, {})
+  end
+
+  local notified = false
+
+  function vim._truncated_echo_once(msg)
+    if not notified then
+      truncated_echo(msg)
+      notified = true
+      return true
+    end
+    return false
+  end
+end
+
+local function traceback()
+  local level = 4
+  local backtrace = { 'stack traceback:' }
+  while true do
+    local info = debug.getinfo(level, 'Sl')
+    if not info then
+      break
+    end
+    local msg = ('  %s:%s'):format(info.source:sub(2), info.currentline)
+    table.insert(backtrace, msg)
+    level = level + 1
+  end
+  return table.concat(backtrace, '\n')
+end
+
+--- Shows a deprecation message to the user. Internal function customized to fit neovim's needs.
+---
+---@param name        string     Deprecated feature (function, API, etc.).
+---@param version     string     Version when the deprecated function will be removed.
+---@param opts        table|nil  Optional parameters.
+---                              - alternative: suggested replacement function
+---
+---@return string|nil # Deprecated message, or nil if no message was shown.
+function vim._deprecate(name, version, opts)
```

Comment:
Changing signature to avoid `nil` positional arguments would be a good idea, one thing to think about the future: do we plan to eventually change the API of `vim.deprecate` match this (or supporting both signatures for a while ~~with deprecating `vim.deprecate`~~)?

If not for any reason, we should use a similar signature and parameters.

Reply:
> Changing signature to avoid `nil` positional arguments would be a good idea, one thing to think about the future: do we plan to eventually change the API of `vim.deprecate` match this (or supporting both signatures for a while ~~with deprecating `vim.deprecate`~~)?

Not a definitive plan, but a strong possibility.